File: java/IEC61970/Dynamics/StandardModels/UnderexcitationLimiterDynamics/UnderexcLimIEEE1.java
package IEC61970.Dynamics.StandardModels.UnderexcitationLimiterDynamics;

import IEC61970.Base.Domain.PU;
import IEC61970.Base.Domain.Seconds;

/**
 * The class represents the Type UEL1 model which has a circular limit boundary
 * when plotted in terms of machine reactive power vs. real power output.
 * 
 * Reference: IEEE UEL1 421.5-2005 Section 10.1.
 * @author pcha006
 * @version 1.0
 * @created 25-Dec-2023 8:32:04 PM
 */
public class UnderexcLimIEEE1 extends UnderexcitationLimiterDynamics {

	/**
	 * UEL center setting (K<sub>UC</sub>).  Typical Value = 1.38.
	 */
	public PU kuc;
	/**
	 * UEL excitation system stabilizer gain (K<sub>UF</sub>).  Typical Value = 3.3.
	 */
	public PU kuf;
	/**
	 * UEL integral gain (K<sub>UI</sub>).  Typical Value = 0.
	 */
	public PU kui;
	/**
	 * UEL proportional gain (K<sub>UL</sub>).  Typical Value = 100.
	 */
	public PU kul;
	/**
	 * UEL radius setting (K<sub>UR</sub>).  Typical Value = 1.95.
	 */
	public PU kur;
	/**
	 * UEL lead time constant (T<sub>U1</sub>).  Typical Value = 0.
	 */
	public Seconds tu1;
	/**
	 * UEL lag time constant (T<sub>U2</sub>).  Typical Value = 0.05.
	 */
	public Seconds tu2;
	/**
	 * UEL lead time constant (T<sub>U3</sub>).  Typical Value = 0.
	 */
	public Seconds tu3;
	/**
	 * UEL lag time constant (T<sub>U4</sub>).  Typical Value = 0.
	 */
	public Seconds tu4;
	/**
	 * UEL maximum limit for operating point phasor magnitude (V<sub>UCMAX</sub>).
	 * Typical Value = 5.8.
	 */
	public PU vucmax;
	/**
	 * UEL integrator output maximum limit (V<sub>UIMAX</sub>).
	 */
	public PU vuimax;
	/**
	 * UEL integrator output minimum limit (V<sub>UIMIN</sub>).
	 */
	public PU vuimin;
	/**
	 * UEL output maximum limit (V<sub>ULMAX</sub>).  Typical Value = 18.
	 */
	public PU vulmax;
	/**
	 * UEL output minimum limit (V<sub>ULMIN</sub>).  Typical Value = -18.
	 */
	public PU vulmin;
	/**
	 * UEL maximum limit for radius phasor magnitude (V<sub>URMAX</sub>).  Typical
	 * Value = 5.8.
	 */
	public PU vurmax;

	public UnderexcLimIEEE1(){

	}
}//end UnderexcLimIEEE1

File: py/IEC61970/Dynamics/StandardModels/UnderexcitationLimiterDynamics/UnderexcLimIEEE1.py
# Converted by an OPENAI API call using model: gpt-3.5-turbo-1106 on Sun Dec 31 17:25:38 2023
from typing import Any, Generic, TypeVar

from IEC61970.Base.Domain.PU import PU
from IEC61970.Base.Domain.Seconds import Seconds
from IEC61970.Dynamics.StandardModels.UnderexcitationLimiterDynamics import UnderexcitationLimiterDynamics


class UnderexcLimIEEE1(UnderexcitationLimiterDynamics):
    """
    The class represents the Type UEL1 model which has a circular limit boundary
    when plotted in terms of machine reactive power vs. real power output.

    Reference: IEEE UEL1 421.5-2005 Section 10.1.
    """

    def __init__(self, kuc=1.38, kuf=3.3, kui=0, kul=100, kur=1.95, tu1=0, tu2=0.05, tu3=0, tu4=0, vucmax=5.8, vuimax=None, vuimin=None, vulmax=18, vulmin=-18, vurmax=5.8):
        super().__init__()
        self.kuc = PU(kuc)  # UEL center setting (K_UC)
        self.kuf = PU(kuf)  # UEL excitation system stabilizer gain (K_UF)
        self.kui = PU(kui)  # UEL integral gain (K_UI)
        self.kul = PU(kul)  # UEL proportional gain (K_UL)
        self.kur = PU(kur)  # UEL radius setting (K_UR)
        self.tu1 = Seconds(tu1)  # UEL lead time constant (T_U1)
        self.tu2 = Seconds(tu2)  # UEL lag time constant (T_U2)
        self.tu3 = Seconds(tu3)  # UEL lead time constant (T_U3)
        self.tu4 = Seconds(tu4)  # UEL lag time constant (T_U4)
        self.vucmax = PU(vucmax)  # UEL maximum limit for operating point phasor magnitude (V_UCMAX)
        self.vuimax = PU(vuimax)  # UEL integrator output maximum limit (V_UIMAX)
        self.vuimin = PU(vuimin)  # UEL integrator output minimum limit (V_UIMIN)
        self.vulmax = PU(vulmax)  # UEL output maximum limit (V_ULMAX)
        self.vulmin = PU(vulmin)  # UEL output minimum limit (V_ULMIN)
        self.vurmax = PU(vurmax)  # UEL maximum limit for radius phasor magnitude (V_URMAX)
